### ISSUE TITLE
Can we exploit build cache for faster launch of R build pack?

### ISSUE THREAD

**Issue by brooksambrose** [jupyterhub/repo2docker#273]
**tldr** for R build pack, any hope of using `postBuild` to add a layer without invalidating cache for `install.R`?

This is a repo2docker question for a binderhub: is there a way to exploit Docker build cache to speed up building images using the r2d R build pack? The use case is a heavy base image with some light user customization on top, e.g. building R with an `install.R` script for tidyverse seems to take about 15 minutes on a beefyish rig. Using binderhub in a course, I want to give students a good data science image and give them the option to layer on top for original student projects, without having to wait for tidyverse to rebuild as well. As far as I can tell there's no way to add a new layer, but want to ask here in case I missed it.

So classic 80/20 problem, possibly same conversation here #115.  Goal is to build a "never variable" heavy layer once, and then give users the option to build a "sometimes variable" lighter layer of  package installs on that, but before the "always variable" data layer.

I'm trying to use the `postBuild` option to add that light layer of R package installs on at the very end, which seems like it would make sense given that the r2d Dockerfile adds a layer just for postBuild as the last line. To the contrary, the behavior seems to be to invalidate the cache ~after using runtime.txt but before using install.R~ before copying anything from the repo, even if most of the repo is the same.

For example, using debug on a local r2d:

```
jupyter-repo2docker --debug --no-build https://github.com/w201rdada/final
```
Gives something like this:

```Dockerfile
#...truncated jupyter build layers!
# Copy and chown stuff. This doubles the size of the repo, because
# you can't actually copy as USER, only as root! Thanks, Docker!
USER root
COPY src/ ${HOME}
RUN chown -R ${NB_USER}:${NB_USER} ${HOME}

# Run assemble scripts! These will actually build the specification
# in the repository into the image.
RUN echo "options(repos = c(CRAN='https://mran.microsoft.com/snapshot/2018-03-19'), download.file.method = 'libcurl')" > /etc/R/Rprofile.site

USER ${NB_USER}
RUN Rscript install.R


# Container image Labels!
# Put these at the end, since we don't want to rebuild everything
# when these change! Did I mention I hate Dockerfile cache semantics?


# We always want containers to run as non-root
USER ${NB_USER}

RUN ./postBuild
```

The behavior seems to be to use cache up until copying the binder files from the repo (line 35), then all cache after is invalidated even if only one file has changed in the repo.

```
Step 31/47 : USER ${NB_USER}
 ---> Using cache
 ---> d7dd9a31bd5c
Step 32/47 : RUN pip install --no-cache-dir nbrsessionproxy==0.6.1 && jupyter serverextension enable nbrsessionproxy --sys-prefix && jupyter nbextension install --py nbrsessionproxy --sys-prefix && jupyter nbextension enable --py nbrsessionproxy --sys-prefix
 ---> Using cache
 ---> 048b05c53777
Step 33/47 : RUN R --quiet -e "install.packages('devtools', repos='https://mran.microsoft.com/snapshot/2018-02-01', method='libcurl')" && R --quiet -e "devtools::install_github('IRkernel/IRkernel', ref='0.8.11')" && R --quiet -e "IRkernel::installspec(prefix='$NB_PYTHON_PREFIX')"
 ---> Using cache
 ---> 55ed3a5f84be
Step 34/47 : USER root
 ---> Using cache
 ---> 5ff918525f3f
Step 35/47 : COPY src/ ${HOME}
 ---> bcf436f92f32
Step 36/47 : RUN chown -R ${NB_USER}:${NB_USER} ${HOME}
 ---> Running in d3f8dc3de618
Removing intermediate container d3f8dc3de618
 ---> 9ed4f812da63
```

**Comment by brooksambrose** [comment]
Also fine if the best approach is to use r2d to build the heavy image then use a Dockerfile in a repo to layer something on top of the image. I wanted to expose the details of the R build pack to students, but having an R installation RUN line in a Dockerfile is a good solution too. Downside of that approach is that the custom package install would be rebuilt every time any other user data changes in the repo, but at least it prevents rebuilding the heavy layer.

**Comment by betatim** [comment]
This is related to how docker layers work. Once one layer has changed all layers after it can not come from the cache any more.

Is there a way to install "binary" packages in R (akin to python wheels or conda packages)?

An alternative would be to build a base image, then run `repo2docker --debug --no-run`, fish out the commands that you need for copying in the repo contents and running `postBuild` and make a `Dockerfile` based on the base image and these commands. This way you get a `Dockerfile` that is compatible/equivalent-ish to what repo2docker does so you can probably keep using `repo2docker` to run things/have mybinder.org compatibility.

**Comment by brooksambrose** [comment]
Yeah I'm not sure about binaries or speeding up the R build itself, though here is an approach: http://www.rexamine.com/2015/07/speeding-up-r-package-installation-process/

@betatim That layering can be controlled though right? So putting a layer that changes a lot before a layer that doesn't change a lot leads to unnecessary building. So again the COPY line below invalidates everything after it. Alternatively the config files and the instructions they generate could be written into the Dockerfile before the whole repo is copied, in an order understood by the user, and if those RUN commands didn't change then their results would come from cache and not be rebuilt. The config files could even then be deleted before the repo is copied to keep it clean.

So in the abstract:

```
# Don't copy and chown stuff yet.
# Run assemble scripts! These will actually build the specification
# in the repository into the image.
RUN echo "options(repos = c(CRAN='https://mran.microsoft.com/snapshot/2018-03-19'), download.file.method = 'libcurl')" > /etc/R/Rprofile.site

# arbitrary layer (use it as a heavy or invariable layer)
RUN R --quiet -e "CATTED CONTENTS OF INSTALL1.R"

# arbitrary layer (use it as a light or variable layer)
RUN R --quiet -e "CATTED CONTENTS OF INSTALL2.R"

# copy and chown stuff
USER root
COPY src/ ${HOME}
RUN chown -R ${NB_USER}:${NB_USER} ${HOME}

USER ${NB_USER}

RUN ./postBuild
```
Then allowing multiple configs of the same type could give the user control over layering, which could then exploit build cache and allow user customization with speedy building of light layers.

**Comment by betatim** [comment]
The idea of copying in the assembly and build scripts, building, copy in rest of the repository is interesting. What is tricky is that some of the assemble/build files can depend on "arbitrary" stuff in the repository. This is super hard to detect.

So we could treat build instructions that can not possibly depend on anything in the repo.

**Comment by manics** [comment]
It's been a while and there are other discussions about optimising R, so I'll close this.
